In [1]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
import seaborn as sns
plt.style.use('ggplot')
# pd.set_option('max_columns', 200)

In [2]:
file_path = r"/Users/danielaina/Documents/Python Work/common_player_info.csv"
df = pd.read_csv(file_path)
print(df.head())


   person_id first_name     last_name   display_first_last  \
0      76001       Alaa     Abdelnaby       Alaa Abdelnaby   
1      76003     Kareem  Abdul-Jabbar  Kareem Abdul-Jabbar   
2       1505      Tariq   Abdul-Wahad    Tariq Abdul-Wahad   
3        949    Shareef   Abdur-Rahim  Shareef Abdur-Rahim   
4      76005        Tom     Abernethy        Tom Abernethy   

  display_last_comma_first  display_fi_last          player_slug  \
0          Abdelnaby, Alaa     A. Abdelnaby       alaa-abdelnaby   
1     Abdul-Jabbar, Kareem  K. Abdul-Jabbar  kareem-abdul-jabbar   
2       Abdul-Wahad, Tariq   T. Abdul-Wahad    tariq-abdul-wahad   
3     Abdur-Rahim, Shareef   S. Abdur-Rahim  shareef-abdur-rahim   
4           Abernethy, Tom     T. Abernethy        tom-abernethy   

             birthdate          school country  ...  \
0  1968-06-24 00:00:00            Duke     USA  ...   
1  1947-04-16 00:00:00            UCLA     USA  ...   
2  1974-11-03 00:00:00  San Jose State  France  ...  

In [3]:
df.shape
# 4171 rows and 33 columns

(4171, 33)

In [4]:
df.head(20)

#showing the first 20 rows

person_id first_name     last_name   display_first_last  \
0       76001       Alaa     Abdelnaby       Alaa Abdelnaby   
1       76003     Kareem  Abdul-Jabbar  Kareem Abdul-Jabbar   
2        1505      Tariq   Abdul-Wahad    Tariq Abdul-Wahad   
3         949    Shareef   Abdur-Rahim  Shareef Abdur-Rahim   
4       76005        Tom     Abernethy        Tom Abernethy   
5       76006     Forest          Able          Forest Able   
6       76007       John     Abramovic       John Abramovic   
7      203518       Alex       Abrines         Alex Abrines   
8     1630173   Precious       Achiuwa     Precious Achiuwa   
9      101165       Alex         Acker           Alex Acker   
10      76008     Donald      Ackerman      Donald Ackerman   
11      76009       Mark         Acres           Mark Acres   
12      76010    Charles         Acton        Charles Acton   
13     203112     Quincy           Acy           Quincy Acy   
14      76011      Alvan         Adams          Alvan Adams   
15      76012        Don         Adams            Don Adams   
16     200801     Hassan         Adams         Hassan Adams   
17    1629121     Jaylen         Adams         Jaylen Adams   
18     203919     Jordan         Adams         Jordan Adams   
19        149    Michael         Adams        Michael Adams   

   display_last_comma_first  display_fi_last          player_slug  \
0           Abdelnaby, Alaa     A. Abdelnaby       alaa-abdelnaby   
1      Abdul-Jabbar, Kareem  K. Abdul-Jabbar  kareem-abdul-jabbar   
2        Abdul-Wahad, Tariq   T. Abdul-Wahad    tariq-abdul-wahad   
3      Abdur-Rahim, Shareef   S. Abdur-Rahim  shareef-abdur-rahim   
4            Abernethy, Tom     T. Abernethy        tom-abernethy   
5              Able, Forest          F. Able          forest-able   
6           Abramovic, John     J. Abramovic       john-abramovic   
7             Abrines, Alex       A. Abrines         alex-abrines   
8         Achiuwa, Precious       P. Achiuwa     precious-achiuwa   
9               Acker, Alex         A. Acker           alex-acker   
10         Ackerman, Donald      D. Ackerman      donald-ackerman   
11              Acres, Mark         M. Acres           mark-acres   
12           Acton, Charles         C. Acton        charles-acton   
13              Acy, Quincy           Q. Acy           quincy-acy   
14             Adams, Alvan         A. Adams          alvan-adams   
15               Adams, Don         D. Adams            don-adams   
16            Adams, Hassan         H. Adams         hassan-adams   
17            Adams, Jaylen         J. Adams         jaylen-adams   
18            Adams, Jordan         J. Adams         jordan-adams   
19           Adams, Michael         M. Adams        michael-adams   

              birthdate                school  country  ...  \
0   1968-06-24 00:00:00                  Duke      USA  ...   
1   1947-04-16 00:00:00                  UCLA      USA  ...   
2   1974-11-03 00:00:00        San Jose State   France  ...   
3   1976-12-11 00:00:00            California      USA  ...   
4   1954-05-06 00:00:00               Indiana      USA  ...   
5   1932-07-27 00:00:00      Western Kentucky      USA  ...   
6   1919-02-09 00:00:00                 Salem      USA  ...   
7   1993-08-01 00:00:00          FC Barcelona    Spain  ...   
8   1999-09-19 00:00:00               Memphis  Nigeria  ...   
9   1983-01-21 00:00:00            Pepperdine      USA  ...   
10  1930-09-04 00:00:00  Long Island-Brooklyn      USA  ...   
11  1962-11-15 00:00:00          Oral Roberts      USA  ...   
12  1942-01-11 00:00:00             Hillsdale      USA  ...   
13  1990-10-06 00:00:00                Baylor      USA  ...   
14  1954-07-19 00:00:00              Oklahoma      USA  ...   
15  1947-11-27 00:00:00          Northwestern      USA  ...   
16  1984-06-20 00:00:00               Arizona      USA  ...   
17  1996-05-04 00:00:00       St. Bonaventure      USA  ...   
18  1994-07-08 00:00:00          

In [5]:
# listing out all the columns

df.columns

Index(['person_id', 'first_name', 'last_name', 'display_first_last',
       'display_last_comma_first', 'display_fi_last', 'player_slug',
       'birthdate', 'school', 'country', 'last_affiliation', 'height',
       'weight', 'season_exp', 'jersey', 'position', 'rosterstatus',
       'games_played_current_season_flag', 'team_id', 'team_name',
       'team_abbreviation', 'team_code', 'team_city', 'playercode',
       'from_year', 'to_year', 'dleague_flag', 'nba_flag', 'games_played_flag',
       'draft_year', 'draft_round', 'draft_number', 'greatest_75_flag'],
      dtype='object')

In [6]:
df.dtypes
# shows what type of data we have in the table columns

person_id                             int64
first_name                           object
last_name                            object
display_first_last                   object
display_last_comma_first             object
display_fi_last                      object
player_slug                          object
birthdate                            object
school                               object
country                              object
last_affiliation                     object
height                               object
weight                              float64
season_exp                          float64
jersey                               object
position                             object
rosterstatus                         object
games_played_current_season_flag     object
team_id                               int64
team_name                            object
team_abbreviation                    object
team_code                            object
team_city                       

In [7]:
df.describe()
# shows us some info and stats of the numeric information in our data set

person_id       weight   season_exp       team_id    from_year  \
count  4.171000e+03  4071.000000  4171.000000  4.171000e+03  4156.000000   
mean   3.327509e+05   211.133382     5.195157  1.339538e+09  1989.303176   
std    5.768886e+05    26.926082     4.470879  6.026618e+08    22.812298   
min    2.000000e+00   133.000000     0.000000  0.000000e+00  1946.000000   
25%    7.612850e+04   190.000000     2.000000  1.610613e+09  1974.000000   
50%    7.759300e+04   210.000000     3.000000  1.610613e+09  1992.000000   
75%    2.022745e+05   230.000000     8.000000  1.610613e+09  2010.000000   
max    1.631347e+06   360.000000    22.000000  1.610613e+09  2022.000000   

           to_year  
count  4156.000000  
mean   1993.746631  
std      23.636712  
min    1946.000000  
25%    1977.000000  
50%    1998.000000  
75%    2015.000000  
max    2023.000000

In [8]:
# cleaning data to drop any columns or information we might not need

 # df.head() # to view columns again
df = df[[
    # 'person_id', 'first_name', 'last_name', 
    'display_first_last',
      # 'display_last_comma_first', 'display_fi_last', 'player_slug',
       'birthdate', 'school', 'country', 
    #'last_affiliation', 
    'height',
       'weight', 'season_exp', 'jersey', 'position', 
    #'rosterstatus',
      # 'games_played_current_season_flag', 'team_id', 
    'team_name',
      # 'team_abbreviation', 'team_code', 'team_city', 'playercode',
       'from_year', 'to_year', 
   # 'dleague_flag', 'nba_flag', 'games_played_flag',
       'draft_year', 'draft_round', 'draft_number', 'greatest_75_flag']].copy() # pull in all columns then copy the columns  and input them withing the two brackets , then from there delte whatever columns you dont want

# hashtagging the columns I dont want 
# add .copy() at the end and create a new df and save df to a new df using df = df with .copy() at the end of subsetting list of column names


In [9]:
# using to df.drop to drop columns

# df.drop(['country'])

In [10]:
df.shape
df.dtypes

display_first_last     object
birthdate              object
school                 object
country                object
height                 object
weight                float64
season_exp            float64
jersey                 object
position               object
team_name              object
from_year             float64
to_year               float64
draft_year             object
draft_round            object
draft_number           object
greatest_75_flag       object
dtype: object

In [11]:
# changing objects to date time columns
df['birthdate'] = pd.to_datetime(df['birthdate'])

# Convert draft_year to datetime
df['draft_year'] = df['draft_year'].apply(lambda x: pd.to_datetime(x, format='%Y', errors='coerce') if x != 'undrafted' else pd.NaT)

# Convert draft_number to numeric, coercing errors to handle non-numeric values
df['draft_number'] = pd.to_numeric(df['draft_number'], errors='coerce')

# Convert draft_round to numeric, coercing errors to handle non-numeric values
df['draft_round'] = pd.to_numeric(df['draft_round'], errors='coerce')



In [12]:
df.dtypes

display_first_last            object
birthdate             datetime64[ns]
school                        object
country                       object
height                        object
weight                       float64
season_exp                   float64
jersey                        object
position                      object
team_name                     object
from_year                    float64
to_year                      float64
draft_year            datetime64[ns]
draft_round                  float64
draft_number                 float64
greatest_75_flag              object
dtype: object

In [13]:
# renaming columns
df = df.rename(columns={'display_first_last': 'First_Last','greatest_75_flag':'Top_75_Greatest','season_exp':'Seasons_played'})
df.head()

First_Last  birthdate          school country height  weight  \
0       Alaa Abdelnaby 1968-06-24            Duke     USA   6-10   240.0   
1  Kareem Abdul-Jabbar 1947-04-16            UCLA     USA    7-2   225.0   
2    Tariq Abdul-Wahad 1974-11-03  San Jose State  France    6-6   235.0   
3  Shareef Abdur-Rahim 1976-12-11      California     USA    6-9   245.0   
4        Tom Abernethy 1954-05-06         Indiana     USA    6-7   220.0   

   Seasons_played jersey       position      team_name  from_year  to_year  \
0             5.0     30        Forward  Trail Blazers     1990.0   1994.0   
1            20.0     33         Center         Lakers     1969.0   1988.0   
2             7.0      9  Forward-Guard          Kings     1997.0   2003.0   
3            13.0      3        Forward      Grizzlies     1996.0   2007.0   
4             5.0      5        Forward       Warriors     1976.0   1980.0   

  draft_year  draft_round  draft_number Top_75_Greatest  
0 1990-01-01          1.0          25.0               N  
1 1969-01-01          1.0           1.0               Y  
2 1997-01-01          1.0          11.0               N  
3 1996-01-01          1.0           3.0               N  
4 1976-01-01          3.0          43.0               N

In [14]:
# fixing height column

# Function to convert height from "feet-inches" to inches
def height_to_inches(height):
    if isinstance(height, str) and '-' in height:
        feet, inches = height.split('-')
        return int(feet) * 12 + int(inches)
    else:
        return np.nan  # Return NaN for missing or incorrect format

# Apply the conversion to the height column
df['height_inches'] = df['height'].apply(height_to_inches)

# Optional: Convert inches to decimal feet (e.g., 6 feet 10 inches = 6.83 feet)
df['height_feet'] = df['height_inches'] / 12

# Display the updated DataFrame
df[['First_Last', 'height', 'height_inches', 'height_feet']]

First_Last height  height_inches  height_feet
0          Alaa Abdelnaby   6-10           82.0     6.833333
1     Kareem Abdul-Jabbar    7-2           86.0     7.166667
2       Tariq Abdul-Wahad    6-6           78.0     6.500000
3     Shareef Abdur-Rahim    6-9           81.0     6.750000
4           Tom Abernethy    6-7           79.0     6.583333
...                   ...    ...            ...          ...
4166          Paul Zipser    6-8           80.0     6.666667
4167           Ante Zizic   6-10           82.0     6.833333
4168             Jim Zoet    7-1           85.0     7.083333
4169          Ivica Zubac    7-0           84.0     7.000000
4170           Matt Zunic    6-3           75.0     6.250000

[4171 rows x 4 columns]

In [15]:
df.dtypes

First_Last                 object
birthdate          datetime64[ns]
school                     object
country                    object
height                     object
weight                    float64
Seasons_played            float64
jersey                     object
position                   object
team_name                  object
from_year                 float64
to_year                   float64
draft_year         datetime64[ns]
draft_round               float64
draft_number              float64
Top_75_Greatest            object
height_inches             float64
height_feet               float64
dtype: object

In [22]:
# identidying missing or null values

df.isna().sum()

First_Last            0
birthdate             0
school               15
country               1
height               96
weight              100
Seasons_played        0
jersey              980
position             63
team_name           702
from_year            15
to_year              15
draft_year         1081
draft_round        1245
draft_number       1304
Top_75_Greatest       0
height_inches        96
height_feet          96
dtype: int64

In [16]:
# finding duplicated data

df.loc[df.duplicated()]

Empty DataFrame
Columns: [First_Last, birthdate, school, country, height, weight, Seasons_played, jersey, position, team_name, from_year, to_year, draft_year, draft_round, draft_number, Top_75_Greatest, height_inches, height_feet]
Index: []

In [23]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.multioutput import MultiOutputRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [28]:
# 1. Convert 'draft_year' to datetime
df['draft_year'] = pd.to_datetime(df['draft_year'], errors='coerce')

# 2. Extract the year
df['draft_year_numeric'] = df['draft_year'].dt.year

# 3. Convert other columns to numeric
df['draft_round'] = pd.to_numeric(df['draft_round'], errors='coerce')
df['height_feet'] = pd.to_numeric(df['height_feet'], errors='coerce')
df['height_inches'] = pd.to_numeric(df['height_inches'], errors='coerce')

# 4. Drop rows with any NaN values in the relevant columns
df_clean = df.dropna(subset=['height_feet', 'height_inches', 'draft_year_numeric', 'draft_round'])


In [29]:
# 5. Features and Targets
X = df_clean[['height_feet', 'height_inches']]
y = df_clean[['draft_year_numeric', 'draft_round']]

# 6. Train/Test Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 7. Model
regr = MultiOutputRegressor(RandomForestRegressor(random_state=42))
regr.fit(X_train, y_train)


MultiOutputRegressor(estimator=RandomForestRegressor(random_state=42))

In [30]:
# 8. Predictions
y_pred = regr.predict(X_test)

# 9. Evaluate
mse = mean_squared_error(y_test, y_pred, multioutput='raw_values')
print(f"Mean Squared Error for Draft Year: {mse[0]}")
print(f"Mean Squared Error for Draft Round: {mse[1]}")

Mean Squared Error for Draft Year: 0.0
Mean Squared Error for Draft Round: 3.7369353721138747


MSE of 3.7369 for draft round means that the model is having trouble predicting the draft round. this could be to weak correlation or the relationship between height and draft round may be more complex than i orginally thought

In [37]:
# predicting seasons played based off of heigh, weight, draft year and round

# Select Features and Target
X = df_clean[['draft_year_numeric', 'draft_round', 'height_inches', 'weight']]
y = df_clean['Seasons_played']

# Train/Test Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Model
model = RandomForestRegressor(random_state=42)
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

# Evaluate
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse}")

Mean Squared Error: 22.73529140602987


In [38]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.model_selection import GridSearchCV

# Define the model
gbr = GradientBoostingRegressor(random_state=42)

# Define hyperparameters to tune
param_grid = {
    'n_estimators': [100, 200, 300],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 5, 7]
}

# Grid search
grid_search = GridSearchCV(estimator=gbr, param_grid=param_grid, cv=5, scoring='neg_mean_squared_error')
grid_search.fit(X_train, y_train)

# Best model
best_gbr = grid_search.best_estimator_

# Predictions
y_pred = best_gbr.predict(X_test)

# Evaluate
mse = mean_squared_error(y_test, y_pred)
print(f"Best Mean Squared Error: {mse}")


Best Mean Squared Error: 17.884373106394495
